server/chat-plugins/friends.ts

```diff
@@ -0,0 +1,623 @@
+				this.database.exec(FS('databases/schemas/friends.sql').readSync());
+			} else {
+				throw e;
+			}
+		}
+		// preparing these once is apparently faster
+		this.statements = {
+			add: this.database.prepare(`REPLACE INTO friends (userid, friend) VALUES(?, ?)`),
+			get: this.database.prepare(`SELECT * FROM friends WHERE userid = ? LIMIT ?`),
+			delete: this.database.prepare(`DELETE FROM friend_requests WHERE receiver = ?`),
+			getSent: this.database.prepare(`SELECT receiver FROM friend_requests WHERE sender = ?`),
+			getReceived: this.database.prepare(`SELECT receiver FROM friend_requests WHERE receiver = ?`),
+			deleteRename: this.database.prepare(
+				`UPDATE friends SET previous_name = null, change_date = null WHERE previous_name = ? AND change_date = ?`
+			),
+			insertRequest: this.database.prepare(`INSERT INTO friend_requests(sender, receiver) VALUES(?, ?)`),
+			deleteRequest: this.database.prepare(`DELETE FROM friend_requests WHERE sender = ? AND receiver = ?`),
+			findFriendship: this.database.prepare(
+				`SELECT * FROM friends WHERE (friend = $user1 AND userid = $user2) OR (userid = $user1 AND friend = $user2)`
+			),
+			renameFriend: this.database.prepare(
+				`UPDATE friends SET friend = $newID, previous_name = $oldID, change_date = $date WHERE friend = $oldID`
+			),
+			renameUserid: this.database.prepare(`UPDATE friends SET userid = $newID WHERE userid = $oldID`),
+			login: this.database.prepare(`UPDATE friends SET last_login = ? WHERE friend = ?`),
+			checkLastLogin: this.database.prepare(`SELECT last_login FROM friends WHERE friend = ?`),
+		};
+	}
+	getFriends(user: User) {
+		if (user.friends) return; // only query once per user object
+		const results = this.statements.get.all(user.id, MAX_FRIENDS).map(item => item.friend);
+		const friends = new Set(results);
+		user.friends = friends;
+		return friends;
+	}
+	getRequests(user: User) {
+		const sent: Set<string> = new Set();
+		const received: Set<string> = new Set();
+		if (user.settings.blockFriendRequests) {
+			// delete any pending requests that may have been sent to them while offline and return
+			this.statements.delete.run(user.id);
+			user.friendRequests = null;
+			return;
+		}
+		const sentResults = this.statements.getSent.all(user.id);
+		for (const request of sentResults) {
+			sent.add(request.receiver);
+		}
+		const receivedResults = this.statements.getReceived.all(user.id);
+		for (const request of receivedResults) {
+			received.add(request.sender);
+		}
+		user.friendRequests = {sent, received};
+		return {sent, received};
+	}
+	getRenames() {
+		// this is only run once, on initialization, so it doesn't need to be in statements
+		const results = this.database.prepare(
+			`SELECT friend, previous_name, change_date FROM friends WHERE previous_name IS NOT NULL`
+		).all();
+		for (const result of results) {
+			const {friend, previous_name, change_date} = result;
+			// delete old data after one week
```

Comment:
Why?

Reply:
Because why keep around the name-change data forever?

Reply:
Can you elaborate on the purpose of name change data? And, why only 1 previous name?

Reply:
So that you don't see a new name on your friends list and get confused who it might be. It's only one previous name because you should only need one, most likely. If not, your friends will probably find out anyway.